\documentclass{article}

\usepackage[tmargin=.9in, bmargin=.9in, lmargin=1in, rmargin=1in]{geometry}
\usepackage{bookmark, wrapfig, enumitem, pdflscape, hyphenat}
\usepackage[labelfont=bf]{caption}
\usepackage{lmodern}
\usepackage[sfdefault]{roboto}
\usepackage[T1]{fontenc}

\usepackage{graphicx}
    \makeatletter % tex.stackexchange.com/a/28565
    \setlength{\@fptop}{0pt}
    \setlength{\@fpbot}{0pt plus 1fil}
    \makeatother

\usepackage[absolute, overlay]{textpos}
    \setlength{\TPHorizModule}{1mm}
    \setlength{\TPVertModule}{1mm}

\usepackage{xcolor}
    \definecolor{WCM}{RGB}{172,31,44} % AC1E2C

\usepackage{hyperref} \hypersetup{
    colorlinks=true, linkcolor={blue!65!black},
    citecolor={blue!65!black}, urlcolor={blue!50!black},
    pdfpagelayout=OneColumn, pdfstartview={XYZ null null 1.25},
    bookmarksnumbered=true, bookmarksopen=true, bookmarksopenlevel=3
}

\usepackage[backend=bibtex, style=nature]{biblatex}
    \addbibresource{latex/references.bib}

\newcommand{\beginsupplement}{
% bytesizebio.net/2013/03/11/adding-supplementary-tables-and-figures-in-latex
    \newpage
    \setcounter{page}{1}
    \renewcommand{\thepage}{S-\arabic{page}}
    \setcounter{table}{0}
    \renewcommand{\thetable}{S\arabic{table}}
    \setcounter{figure}{0}
    \renewcommand{\thefigure}{S\arabic{figure}}
 }

\usepackage{setspace}

\begin{document}

\begin{center}
    \Large{\textbf{Haplotype Diversity and Sequence Heterogeneity of Human Telomeres}}
    \\~\\
    \small{
        Kirill Grigorev\textsuperscript{1,2 \#},
        Jonathan Foox\textsuperscript{1,2,3 \#},
        Daniela Bezdan\textsuperscript{1,2,3},
        Daniel Butler\textsuperscript{1},
        Jared J. Luxton\textsuperscript{4,5},
        Jake Reed\textsuperscript{1},
        \\%rem
        Miles J. McKenna\textsuperscript{4,5},
        Lynn Taylor\textsuperscript{4,5},
        Kerry A. George\textsuperscript{4,5},
        Cem Meydan\textsuperscript{1,2,3},
        Susan M. Bailey\textsuperscript{4,5 *},
        Christopher E. Mason\textsuperscript{1,2,3,6 *}
    }
\end{center}

\small{ \noindent
    \textsuperscript{1} Department of Physiology and Biophysics, Weill Cornell Medicine, New York, New York, USA
    \\
    \textsuperscript{2} The HRH Prince Alwaleed Bin Talal Bin Abdulaziz Alsaud Institute for Computational Biomedicine, \\
    \textcolor{white}{\textsuperscript{2}} Weill Cornell Medicine, New York, New York, USA
    \\
    \textsuperscript{3} The Feil Family Brain and Mind Research Institute, New York, New York, USA
    \\
    \textsuperscript{4} Department of Environmental and Radiological Health Sciences, Colorado State University, Fort Collins, CO
    \\
    \textsuperscript{5} Cell and Molecular Biology Program, Colorado State University, Fort Collins, CO
    \\
    \textsuperscript{6} The WorldQuant Initiative for Quantitative Prediction, Weill Cornell Medicine, New York, NY, USA
    \\
    \textsuperscript{\#} Co-first authors
    \\
    \textsuperscript{*} Corresponding authors. Send correspondence to S.M.B. (susan.bailey@colostate.edu) \\%rem
    \textcolor{white}{\textsuperscript{*}} and C.E.M. (chm2042@med.cornell.edu)
}

\normalsize
\doublespacing

\section*{Abstract} \addcontentsline{toc}{section}{Abstract}
    Telomeres are regions of repetitive nucleotide sequences capping the ends of eukaryotic chromosomes that protect against deterioration,
        and whose lengths can be correlated with age and adverse health risk factors.
    Given their length and repetitive nature,
        telomeric regions are not easily reconstructed from short-read sequencing,
        making telomere sequence resolution a very costly and generally intractable problem.
    Recently, long-read sequencing, with read lengths measuring in hundreds of Kbp,
        has made it possible to routinely read into telomeric regions and inspect their sequence structure.
    Here, we describe a framework for
        extracting telomeric reads from whole genome single-molecule sequencing experiments,
        prior-less \textit{de novo} identification of telomere repeat motifs,
        and describing their sequence variation.
    We find that
        long telomeric stretches can be accurately captured with long-read sequencing,
        observe extensive sequence heterogeneity of human telomeres,
        discover and localize non-canonical motifs (both previously reported as well as novel),
        confirm the presence of the non-canonical motifs in short read sequencing experiments,
        and report the first motif composition maps of human telomeric haplotypes across populations on a multi-Kbp scale.

\section*{Keywords} \addcontentsline{toc}{section}{Keywords}
    Telomere, telomeric haplotypes, long-read sequencing, telomere sequence heterogeneity

\pagebreak
%\singlespacing
%\tableofcontents
\doublespacing

\section*{Introduction} \addcontentsline{toc}{section}{Introduction}
    Telomeres are the functional ends of human chromosomes that naturally shorten with cell division and therefore with age \cite{teloaging}.
    Telomere length can also be influenced by a variety of lifestyle factors and environmental exposures
        (e.g., stress, exercise, air pollution, radiation) \cite{teloeffects}.
    While human telomeres are known to consist largely of a conserved six-nucleotide repeat (TTAGGG) \cite{moyzis},
        several studies have identified variations of this motif in proximal telomeric regions
            \cite{telovars1989,telovars1999,telovars2018,telovars2019}.
    However, such studies were performed with oligonucleotide hybridization, PCR, immunoprecipitation, and short-read sequencing,
        requiring prior assumptions about specific target motifs, custom sample preparation, and targeted sequencing,
        and therefore preventing \textit{de novo} identification of motif variants and their localization.
    Thus, long-range maps of telomeric sequence variation in the human genome are still lacking.
    Such maps can provide insight into telomere biology and enable novel approaches to analyze the effects of
        health status, aging, and environment on telomere sequence and length.
    \\~\\
    To improve our understanding of telomere sequence structure and variation, we developed \textit{edgeCase},
        a framework for alignment and \textit{de novo} telomeric motif discovery
        which uses human whole genome long-read sequencing experiments,
            making it easily scalable.
    We have validated these methods using
        Genome in a Bottle \cite{giab} single-molecule real-time (SMRT) sequencing datasets
            generated with Pacific Biosciences circular consensus sequencing (PacBio CCS) \cite{pacbio,pacbioccs},
        and short-read Illumina \cite{illumina} and 10X Genomics (Chromium) \cite{10x} datasets.
    These results provide evidence for
        multiple novel, non-canonical telomeric repeats,
        resolution of multiple chromosome-specific haplotypes with SMRT sequencing,
        and a new method for long-range characterization of the structure of telomeric sequences.

\section*{Results} \addcontentsline{toc}{section}{Results}

\subsection*{A telomere-annotated reference genome enables recovery of telomeric reads from human long-read whole genome sequencing datasets}
\addcontentsline{toc}{subsection}{A telomere-annotated reference genome enables recovery of telomeric reads from human long-read whole genome sequencing datasets}
    We constructed an extended reference genome, \textit{hg38ext},
        that combines chromosome sequences of the \textit{hg38} reference genome \cite{grch38,hg38}
            and human subtelomeric assemblies \cite{riethman2014},
        resulting in a reference set annotated with boundaries of subtelomeric and telomeric tracts.
            % (see \hyperref[sec:methods]{Materials and Methods}).
        The layout of this reference set is available in \textbf{Supplemental File S1},
            and the set itself can be reproduced with a script available as \textbf{Supplemental File S2}.
    We then aligned PacBio CCS reads of seven Genome in a Bottle (GIAB, \cite{giab}) human subjects (HG001 through HG007) to \textit{hg38ext},
        and in total, observed reads mapping to the ends of chromosomes and extending into telomeric regions
            on 10 \textit{p} arms and 19 \textit{q} arms,
                with 53\textendash{}295 such reads on the \textit{p} arms
                and 384\textendash{}1119 on the \textit{q} arms
                    (\textbf{Supplemental Table S1}).
    Portions of reads contained in the telomeric regions were extracted for further analysis (\textbf{\autoref{fig:schematic_alignment}}).
        \begin{figure}[h!] \centering % \vspace{5mm}
        \includegraphics[height=.75\textheight,width=\textwidth,keepaspectratio]{renders/figures/Figure-1.pdf}
        \caption{
             \small Mapping of candidate telomeric reads, illustrated with reads from the HG002 dataset aligning to \mbox{chromosome 5.}
             The chromosome is displayed schematically, centered around the centromere.
             Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
             Coordinates are given in Kbp, relative to the positions of the telomeric tract boundaries.
             Statistics for all chromosomes of all seven datasets are provided in \textbf{Supplemental Table S1}.
        }
        \label{fig:schematic_alignment}
        \end{figure}

\subsection*{Telomeric long reads contain variations of the canonical motif}
\addcontentsline{toc}{subsection}{Telomeric long reads contain variations of the canonical motif}
    We performed \textit{de novo} repeat discovery in the telomeric sequences for motifs of lengths 4 through 16,
            and identified motifs in repeat contexts that are statistically enriched in the seven datasets.
        The majority of motifs were either the canonical TTAGGG / CCCTAA,
            its variations (e.g., TT\underline{G}GGG / CCC\underline{C}AA),
            or a duplet of variants, such as TTAGGGTTA\underline{G}GGG
            (\textbf{\autoref{tab:repeatfinder}}).
        CG-rich motifs were also observed on the \textit{p} arms.
        The top enriched motif (TTAGGG / CCCTAA) explained
            62.2\%\textendash{}82.5\% of the telomeric repeat content on the \textit{q} arms
            and 11.6\%\textendash{}36.3\% on the \textit{p} arms,
            and three more motifs (TT\underline{G}GGG, TTA\underline{G}GGG, T\underline{G}AGGG)
                each explained at least 1\% of the repeat content in all seven datasets.
    \\~\\
    We visualized the locations of the top four enriched motifs and their reverse complements
        on the chromosomal ends of the HG002 dataset (\textbf{\autoref{fig:hg002_densityplot}}),
        as it provided the deepest coverage among the assessed datasets (\textbf{Supplemental Table S1}).
        Only the chromosomal arms covered by at least 25 reads were plotted.
            Plots for the other six datasets are available as \textbf{Supplemental Figs. S1\textendash{}S6}.
        \input{latex/tables/Table-1-repeatfinder.tex} % tab:repeatfinder
        \begin{figure}[h!] \centering
        \includegraphics[height=.88\textheight,width=\textwidth,keepaspectratio]{renders/figures/Figure-2.pdf}
        \caption{
             \small Densities of the top four enriched motifs at ends of chromosomal \textbf{(A)} \textit{p} arms
             and \textbf{(B)} \textit{q} arms of the HG002 dataset.
             \textit{Background} represents the remaining sequence content (non-repeating sequence and not significantly enriched motifs).
             Only the arms covered by at least 25 reads are displayed.
             Reads are shown aligned to the contigs in the \textit{hg38ext} reference set, and genomic coordinates are given in Kbp.
             Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
        }
        \label{fig:hg002_densityplot}
        \end{figure}
    \noindent % \\~\\
    Long reads on each arm agreed on the locations of different motifs within any given 10 bp window
        (the coverage-weighted median of normalized Shannon entropy was 0.00 for all data,
        and the coverage-weighted 3rd quartile was 0.00\textendash{}0.13,
        \textbf{\autoref{fig:entropy}}),
        indicating that locations of the variations are colinear among reads.
        \begin{figure}[h!] \centering
        \includegraphics[height=.75\textheight,width=\textwidth,keepaspectratio]{renders/figures/Figure-3.pdf}
        \caption{
            \small Distribution of motif entropies in 10 bp windows of of candidate PacBio CCS reads aligning to the same chromosomal arms
            in GIAB datasets HG001 through HG007,
            with respect to per-window coverage,
            and the coverage-weighted quartiles of the entropy values.
            % no fourth quartile: https://math.stackexchange.com/a/1468111
        }
        \label{fig:entropy}
        \end{figure}

\subsection*{Short-read sequencing validates motif variations observed in long reads}
\addcontentsline{toc}{subsection}{Short-read sequencing validates motif variations observed in long reads}
    We validated these findings using short-read sequencing in two ways.
    First, we found that telomeric candidate reads extracted with \textit{Telomerecat} \cite{telomerecat}
        from respective GIAB Illumina datasets
        supported a definitive majority of the telomeric sequence on the \textit{q} arms of long-read datasets
            (median 98\% of the sequence supported, \textbf{Supplemental Fig. S7}).
        At the same time, only a median of 38\% of the sequence of the \textit{p} arm long reads was supported,
            and while this does not automatically invalidate them (see \hyperref[sec:discussion]{Discussion}),
            only the long reads mapping to the \textit{q} arms were used for downstream analyses.
    Second, we confirmed 15 enriched motifs in independently generated human short-read and linked-read genomic datasets,
        with the same four motifs being the most enriched (\textbf{Supplemental Table S2}).

\subsection*{Long-read sequencing resolves human telomeric haplotypes} %TODO uncovers a variety
\addcontentsline{toc}{subsection}{Long-read sequencing resolves human telomeric haplotypes}
%TODO discuss causes of entropy: seq. errors, SV, haplotypes
    While reads agreed on colinearity of motifs, evidenced by low entropy,
        rarer non-zero entropy values could be attributable both to sequencing errors
        and to structural variations within the same subject's dataset.
    To investigate this possibility,
        we clustered reads on each \textit{q} arm of each subject by relative pairwise Levenshtein distances \cite{levenshtein}
        and found that hierarchical clustering described read similarity well,
            resulting in high cophenetic correlation between the dendrograms and the pairwise distance matrices
                (\textbf{\autoref{tab:cophenetic}}),
            and in visible structure (\textbf{\autoref{fig:haplotypes}}).
            % \eject \pdfpageheight=40in \pdfpagewidth=22in
                \begin{figure}[h!] \centering % \vspace{5mm}
                \includegraphics[height=.88\textheight,width=\textwidth,keepaspectratio]{renders/figures/Figure-4-lowdpi.png}
                \caption{
                    \small Clustering of reads by relative pairwise Levenshtein distances (unitless measure)
                    on each chromosomal \mbox{\textit{q} arm} of datasets HG001 through HG007,
                    and densities of top four enriched motifs along each read.
                    Each horizontal line represents an individual read;
                    Genomic coordinates are given in Kbp, relative to the positions of the telomeric tract boundaries.
                    Only the chromosomal arms cumulatively covered by at least 25 reads are displayed.
                }
                \label{fig:haplotypes}
                \end{figure}
            % \eject \pdfpageheight=11in \pdfpagewidth=8.5in
            \input{latex/tables/Table-2-cophenetic.tex} % tab:cophenetic
    ~\\
    In this complex clustering, subject- and population-specific variation was evident and quantifiable via relative Levenshtein distances
        (\textbf{\autoref{tab:haptests}}; see \hyperref[sec:methods]{Materials and Methods}):
        overall,
            telomeric reads within a subject were more similar than within a population
                (adjusted Wilcoxon signed-rank test \textit{p} = 7.7e-223),
            and telomeric reads within a population were more similar than between populations
                (\textit{p} = 3.1e-120).
        Importantly, however, this was true for most, but not all reads;
            4\% of all assessed reads (142 out of 3,686) contributed to interpopulation similarity;
                these reads were twice as close to reads from a different population than they were to any reads of their own subjects.
            (Another 388 reads, or 11\%, also clustered away from their respective subjects, but by a less than 2x distance ratio).
            This trend is observable on \textbf{\autoref{fig:haplotypes}},
                with subjects and populations appearing in a mixed order along the clustering.
        Therefore, the captured reads reflected spectra of haplotypes,
            generally describing subject- and population-specific similarities,
            but including a sizable component that described interpopulation similarity.
        Familial inheritance of variation was also observed:
            either parent's telomeric reads were more similar to their son's than to the other parent's reads in the Azhkenazim trio.
            In the Chinese trio,
                the amount of inheritance of variation from either parent to the son was not found to be statistically significant overall,
                but was present on several of the assessed chromosomes (\textbf{Supplemental Table S3}).
        \input{latex/tables/Table-3-haplotype-comparisons.tex}

\section*{Discussion} \addcontentsline{toc}{section}{Discussion} \label{sec:discussion}
% TODO: non-invalidation due to BIAS of telomerecat and who-knows-what-goings-on on the p arm (biology? tech artifacts?)
    Repeat-rich, low-complexity regions of the human genome such as telomeres
        have been historically recalcitrant to full mapping and annotation \cite{miga2015},
        mainly due to the alignment challenge they pose and to the read lengths required to span such areas \cite{ngslowcomplexity}.
    The advent of long-read, single-molecule methods (third generation sequencing)
        has provided new opportunities to map the sequence composition of a previously "dark" area of the human genome,
        enabling research into the sequence composition and length dynamics \cite{luxton2020} of telomeres.
    Our results reaffirm
        that the canonical repeat (TTAGGG) is certainly the most dominant type of motif in telomeres,
        but also reveal a surprising diversity of repeat variations,
            which are confirmed by both short and long-read sequencing technologies.
    This diversity of repeats includes previously reported variants,
        as well as novel motifs that are characterized not only by nucleotide substitutions,
        but also insertions, deletions, and even motif pairing.
    Apart from these variations,
        CG-rich motifs were identified in telomeric regions of \textit{p} arms,
        consistent with previously reported findings \cite{cpg}.
    Moreover, while short read sequencing is able to identify such variants,
        it alone cannot reveal the relative locations of these motifs within telomeres,
        as repetitive short reads can
            neither be aligned outside of the reference genome
            nor provide enough overlap variability to be assembled \textit{de novo}.
    Long SMRT reads, on the other hand,
        can be anchored to known subtelomeric sequences of the human genome and extend into the previously unmapped telomeric area.
    Furthermore, in contrast to previously published research that utilized targeted sequencing
        \cite{telovars1989,telovars1999,telovars2018,telovars2019},
        the method described here allows identification of multiple enriched motifs and their localization
            \textit{de novo}, without any bias introduced by prior knowledge about the sequence of target motifs.
    These results also highlight the need of better subtelomeric and telomeric annotations in the human genome:
    Four of the 40 subtelomeric assemblies \cite{riethman2014}
        were homologous to regions in the reference genome far within the respective chromosomes (up to 586 Kbp into the reference sequence),
        and the canonical motif was present on the \textit{q} arm of chr8
            only after 2\textendash{}3Kbp past the annotated boundary in all datasets,
        suggesting that the existing assemblies do not provide a completely accurate telomeric annotation,
        and that methods described herein could help to resolve these areas of reference genomes.
    \\~\\
    We observed PacBio CCS reads reaching up to 16 Kbp beyond the known regions of the genome,
        and resolving the underlying sequence with reasonable fidelity,
        measured both by the entropy of motif assignment and by pairwise Levenshtein distances
            between the reads belonging to the same chromosomal arms.
    While short reads also provided support for non-canonical motifs,
        the overlap between the short and the long reads was substantial, but not complete,
        which can be explained by the necessary bias towards the canonical motif during the selection of short reads.
    Therefore, telomeric regions with higher content of non-canonical repeats are less likely to be identified through the use of short reads,
        and instead, long reads appear to be more suitable for this purpose as well.
    Of note, the captured PacBio CCS reads that mapped to \textit{q} arm telomeres
        agreed with short read sequences much better (median 98\% support, \textbf{Supplemental Fig. S7})
        than the reads that mapped to \textit{p} arm telomeres (median 38\% support).
        While, for this reason, we opted to choose only the \textit{q} arm reads for deeper analysis,
            this discrepancy can lend itself to multiple potential explanations,
                from artifacts of SMRT technology preventing faithful reproduction of the sequence,
                to the inherent bias of short-read based methods \cite{telomerecat} towards the canonical motif CCCTAA
                    that in fact could be present at a lower percentage on \textit{p} arm telomeres.
            Therefore, more research is required to determine the level of deviation of \textit{p} arm telomeric sequences
                from those on \textit{q} arms, and into biases and limitations of different technologies
                    for sequencing distal \textit{p} arm chromosomal regions.
    \\~\\
    The identified variations in long range contexts elucidate
        subject-specific, trio- and population-specific similarities of \textit{q} arm telomeric sequences,
            as well as a level of interpopulation similarity,
        and thus provide a new means of haplotype mapping and reveal the existence and motif composition of haplotype spectra
            on a multi-Kbp scale.
    Clustering on a per-subject basis concealed interpopulation similarity,
        but underscored intra-subject variation (\textbf{Supplemental Fig. S8}),
            suggesting coexistence of multiple telomeric haplotypes per \textit{q} arm within each subject.
        Given that the reference DNA for the subjects HG001 through HG007
            was extracted from growths of B lymphoblastoid cell lines,
            this suggests that as B cells undergo maturation,
                distinct clones may gain distinct variations in their telomeric sequence.
        This opens up avenues of investigation into the haplotypic variation among not only immune cells,
            but also different cell types overall.
%TODO for reviewers: stress that the shift from diplotypes to spectra does not invalidate the previous findings: we were able to put these findings and the method into a broader context and see more structure.

\section*{Materials and Methods} \addcontentsline{toc}{section}{Materials and Methods} \label{sec:methods}

\subsection*{The extended reference genome}
\addcontentsline{toc}{subsection}{The extended reference genome}
    We constructed the extended reference genome by performing an all-to-all alignment
        of all contigs in the \textit{hg38} reference genome \cite{grch38,hg38}
        and the subtelomeric assemblies \cite{riethman2014}
        with \textit{minimap2} \cite{minimap} using three settings
            for assembly-to-reference mapping (\textit{asm5}, \textit{asm10}, \textit{asm20}).
    Forty subtelomeric contigs mapped to ends of \textit{hg38} chromosomes with a mapping quality of 60,
        one (XpYptel) mapped with the quality of 0 and was discarded;
        one (14qtel) mapped to the ALT version of chr14 (chr14\_KI270846v1\_alt) with the quality of 52,
            which, in turn, mapped to the main chr14 chromosome with the quality of 60.
    These data and the exact match and mismatch coordinates were used to create a combined reference (\textit{hg38ext})
        in which subtelomeric contigs informed the locations of the boundaries of the telomeric tracts (\textit{tract\_anchor}).
    Such contigs that mapped fully within \textit{hg38} chromosomes resulted in \textit{tract\_anchor} annotations
            directly on those \textit{hg38} chromosomes;
        partially mapping contigs were considered as forking from the \textit{hg38} sequence and were similarly annotated by themselves.

\subsection*{Detection of telomeric sequences in long-read datasets}
\addcontentsline{toc}{subsection}{Detection of telomeric sequences in long-read datasets}
    Seven subjects were selected for the analysis.
    The first individual (NA12878/HG001) came from the pilot genome of the HapMap project \cite{HG001},
    while the other six,
        including
            the Ashkenazim Jewish Trio (son: NA24385/HG002, father: NA24149/HG003, mother: NA24143/HG004)
            and the Chinese Trio \\%rem
            (\mbox{son: NA24631/HG005}, father: NA24694/HG006, mother: NA24695/HG007),
        are members of the Personal Genome Project,
            whose genomes are consented for commercial redistribution and reidentification \cite{HG00X}.
    These subjects are referred to hereafter as HG001 through HG007, respectively.
    \\~\\
    Multiple Genome in a Bottle \cite{giab} PacBio CCS \cite{pacbio,pacbioccs} datasets were available and combined per each subject,
        with mean coverages of individual datasets ranging from $\sim$21x to $\sim$69x (\textbf{Supplemental Table S4}).
    Reads were mapped to \textit{hg38ext} with \textit{minimap2},
        and reads that mapped to either end of either chromosome and overlapped the boundary of its telomeric tract
        were selected for further analysis (\textbf{\autoref{fig:schematic_alignment}}).
    These reads had a portion of their sequence mapped to the reference contig and a portion extending beyond the reference
        (soft- or hard-clipped in the alignment file).
    Sequences past the \textit{tract\_anchor} marker were extracted from the reads that had this marker within their mapped portion
        (from the 5' end to the marker on \textit{p} arms and from the marker to the 3' end on \textit{q} arms,
        accounting for forward and reverse mappings).

\subsection*{Evaluation of telomeric content in short- and linked-read datasets}
\addcontentsline{toc}{subsection}{Evaluation of telomeric content in short- and linked-read datasets}
    To evaluate the concordance of telomeric reads captured by long- and short-read technologies,
        we extracted candidate telomeric reads from GIAB Illumina datasets for each subject (\textbf{Supplemental Table S4})
            with \textit{Telomerecat} \cite{telomerecat},
            and mapped the short reads back onto the candidate long reads from the same subject's dataset with \textit{minimap2},
            allowing all secondary mappings.
        Then, we calculated the fractions of each long read that was supported by the short reads that aligned to it.
    \\~\\
    To evaluate sequence motifs in independent samples collected from human subjects (as opposed to reference cell lines),
        we generated four whole-genome Illumina datasets (mean coverage $\sim$104x)
        and three linked-read 10X datasets (mean coverage $\sim$28x) for one individual at different timepoints,
        and one additional linked-read 10X dataset (coverage $\sim$47x) for another individual.
    These data were originally obtained from astronaut subjects for an unrelated space biology experiment,
        and the blood samples were collected from the subjects as described in \cite{twins_study}.
        For each sample, 1.2ng of sorted immune cell input was aliquoted for
            TruSeq PCR-free WGS (short read) and standard Chromium 10X whole genome (linked-read) preparation respectively,
            and sequenced across one S4 flow cell on an Illumina NovaSeq 6000.
    From these datasets, candidate telomeric short reads were selected using Telomerecat \cite{telomerecat}.

\subsection*{Identification of repeat content}
\addcontentsline{toc}{subsection}{Identification of repeat content}
    Overrepresentation of motifs of lengths $k \subset [4 .. 16]$ was tested within the candidate telomeric regions of PacBio CCS reads,
        as well as in the candidate reads from independently generated Illumina and 10X Chromium datasets.
    To target motifs in repeat contexts,
        doubled sequences (for example, \textit{k}-mer ACGTACGT for motif ACGT) were counted with \textit{jellyfish} \cite{jellyfish},
        and counts of \textit{k}-mers synonymous with respect to circular shifts (for example, ACGTACGT and CGTACGTA) were summed together.
    For each such \textit{k}-mer,
        Fisher's exact test was performed to determine whether its count is significant
            on the background of counts of other \textit{k}-mers of the same length.
    Briefly, we considered \textit{k}-mers with counts higher than 1.5 interquartile range above the third quartile of the distribution
        as potentially classifiable,
        and a 2\texttimes{}2 contingency matrix $ C $ for the test was constructed as follows:
        row 0 contained counts of potentially classifiable \textit{k}-mers,
        row 1 contained counts of remaining (non-classifiable) \textit{k}-mers,
        columns 0 and 1 contained counts of single and remaining (background) \textit{k}-mers, respectively,
            i.e.:
                $ C_{0,0} = $ {\rmfamily count of target \textit{k}-mer},
                $ C_{0,1} = $ {\rmfamily sum of counts of other potentially classifiable \textit{k}-mers},
                $ C_{1,0} = $ {\rmfamily median count of \textit{k}-mer},
                $ C_{1,1} = $ {\rmfamily sum of counts of other non-classifiable \textit{k}-mers}.
    The resultant \textit{p}-values for each motif among the samples were combined using the Mudholkar-George method \cite{george}
        within each technology (PacBio CCS, Illumina, 10X Genomics),
        and the Bonferroni multiple testing correction was applied.
    Motifs in the long-read datasets for which \textit{k}-mers yielded \textit{p}-values below the cutoff of 0.05 were reported.
    As even doubled sequences (such as ACGTACGT for motif ACGT) can partially overlap at the boundaries of repeat contexts,
        we quantified their presence in the telomeric reads in two ways.
        Consider a sequence such as TTAGGG(\underline{TTAGTTAG})GGTTA:
            the inner (TTAG)\texttimes{}2 repeat can be explained by the repeats of the canonical motif extending from either side;
            the middle part of a similar sequence with a bigger number of the repeats of the 4-mer, TTAGGG\underline{TTAG(TTAGTTAG)TTAG}GGTTA,
                can only be explained by the repeats of said 4-mer.
        On the one hand, the maximum fraction of the sequence that can be explained by any one motif is a useful metric,
            and it was calculated and reported.
        On the other hand, the fraction of the sequence that is attributable to only one specific motif
            elucidates the extent of deviation from the surrounding repeat context,
            and identifies motifs that most affect the sequence structure;
            it was calculated as well and reported as each motif's score.
    Additionally, motifs that were significantly enriched in the datasets produced by all three technologies (PacBio, Illumina, 10X),
        with respect to reverse-complemented equivalence, were reported.

\subsection*{Evaluation of sequence concordance in telomeric long reads}
\addcontentsline{toc}{subsection}{Evaluation of sequence concordance in telomeric long reads}
    As telomeric reads contain long low-complexity regions and present an alignment challenge,
        we evaluated concordance of their sequences without realignment of their portions that extended past the reference sequence.
    To that end, for all reads mapping to the same chromosomal arm,
        we calculated densities of each identified motif in a rolling window starting from the innermost mapped position of each entire read.
    To evaluate whether the reads on the same arm agree on the positions of different motifs,
        for each read, we calculated motif densities in 10 bp windows with 10 bp smoothing to buffer insertions and deletions.
    For each window in each read,
        the motif with the highest density was selected to represent that window.
    Then, normalized Shannon entropy among all reads was calculated in each window as $ S = \frac{ - \sum_{i} \; ( p_{i} ln p_{i} )}{ln N} $,
        where $ p_{i} $ is the frequency of each motif in the window and $ N $ is the number of motifs \cite{hepc_entropy}.
    The value of normalized entropy was a metric bounded by $ [ 0, 1 ] $,
        with $ 0 $ describing perfect agreement and $ 1 $ describing maximum randomness.
    For motif visualization,
        we performed 1000 rounds of bootstrap of the calculated density values in the 10 bp rolling windows,
            and selected the lower and the upper bounds of the 95\% confidence interval of bootstrap.
    Of note, several chromosome arms had the \textit{tract\_anchor} position further away from the end of the contig than others
            ($\sim$79\textendash{}586 Kbp into the chromosome sequence),
        and the reads mapping to these arms did not contain these motifs,
        suggesting that
            either their subtelomeric annotations were incorrect
            or large insertions or duplications were present in the reference genome;
        in light of this, reads mapping to
            the \mbox{\textit{p} arm} of chr1,
            the \textit{q} arm of chr4,
            and both arms of chr20
        were removed from the study,
        and the analysis was repeated.

\subsection*{Extraction of telomeric haplotypes from long-read datasets} %TODO rewrite whole section to talk about spectra XXX!!
\addcontentsline{toc}{subsection}{Extraction of telomeric haplotypes from long-read datasets}
Within groups of reads mapping to each chromosome arm, all relative pairwise Levenshtein distances were calculated. %TODO explain it's edit dist
In short, to calculate the absolute distance between each pair of reads, the sequences in the overlapping positions of the reads were extracted; the distance then equaled the minimum number of single-character insertions, deletions, and substitutions required to make these sequences identical. %TODO explain calculation a bit more
The relative distance was computed as the absolute distance divided by the length of the overlap.
Relative distances were then clustered using Ward's method via the Euclidean metric.
%TODO write so much here XXX XXX XXX XXX!!!
%The optimal number of clusters was determined by maximizing the Bayesian information criterion \cite{bic}, allowing for no more than one outlier and at least five reads per cluster, and silhouette scores for these clusterings were calculated.
%Briefly, as previously described \cite{silhouette}, a silhouette score of a clustering was computed as the mean value of silhouette coefficients of all entries, which, in turn, equaled $ (b - a) \over{max(a, b)} $ where $ a = $ {\rmfamily mean intra-cluster distance} and $ b = $ {\rmfamily mean nearest-cluster distance} for an entry.
%Levenshtein distances of all within-cluster pairings and of all out-of-cluster pairings were compared using the one-tailed Mann-Whitney U test; \textit{p}-values were adjusted with the Bonferroni correction.
%Distinct clusters of reads within the same chromosome arm (adjusted Mann-Whitney U \textit{p}-value below 0.05) were reported as putative haplotypes.
%As the HG002 dataset was combined from two sequencing experiments, we investigated the provenance of reads in these haplotypes; reads from both sequencing experiments contributed to each haplotype with an average $\sim$1:2 ratio (\autoref{tab:hg002_haplotype_assignment}).

\section*{Data access} \addcontentsline{toc}{section}{Data access}
The NASA Life Sciences Data Archive (LSDA) is the repository for all human and animal research data, including the whole genome Illumina and 10X Chromium sequencing datasets from subjects aboard the ISS that were used in this study.
These datasets are protected by the terms of the Weill Cornell Medicine Internal Review Board (IRB) and can be made available to be shared upon request.
LSDA has a public facing portal where data requests can be initiated (\href{https://lsda.jsc.nasa.gov/Request/dataRequestFAQ}{lsda.jsc.nasa.gov/Request/dataRequestFAQ});
the LSDA team provides the appropriate processes, tools, and secure infrastructure for archival of experimental data and dissemination while complying with applicable rules, regulations, policies, and procedures governing the management and archival of sensitive data and information.
The LSDA team enables data and information dissemination to the public or to authorized personnel either by providing public access to information or via an approved request process for information and data from the LSDA in accordance with NASA Human Research Program and JSC Institutional Review Board direction.
\\~\\
%\section*{Availability and implementation} \addcontentsline{toc}{section}{Availability and implementation}
The software for identification of telomeric reads, \textit{de novo} discovery of repeat motifs, haplotype inference and motif density visualization was implemented in Python and is freely available at \\%rem
\href{https://github.com/lankycyril/edgecase}{github.com/lankycyril/edgecase}.

\section*{Acknowledgements} \addcontentsline{toc}{section}{Acknowledgements}
We would like to thank
the Epigenomics Core Facility at Weill Cornell Medicine,
the Scientific Computing Unit (SCU),
XSEDE Supercomputing Resources,
as well as
the STARR grants I9-A9-071, I13-0052,
The Vallee Foundation,
The WorldQuant Foundation,
The Pershing Square Sohn Cancer Research Alliance,
NASA (NNX14AH51G, NNX14AB02G, NNX17AB26G),
The National Institutes of Health (R01MH117406, \\%rem
R01NS076465, R01CA249054, R01AI151059, P01HD067244, P01CA214274),
TRISH (NNX16AO69A:0107, \\%rem
NNX16AO69A:0061),
the LLS (9238-16, Mak, MCL-982, Chen-Kiang),
and
the NSF (1840275).

\section*{Author contributions} \addcontentsline{toc}{section}{Author contributions}
S.M.B. and C.E.M. conceived the study.
K.G., J.F., and C.E.M. developed the framework and analyzed the data.
D.Bu., J.J.L., M.J.M., L.T., and K.A.G. participated in collection and processing of the ISS samples.
D.Be., D.Bu., J.J.L, J.R., and C.M. analyzed the data.
All authors edited the manuscript.

\section*{Competing interests} \addcontentsline{toc}{section}{Competing interests}
The authors declare no relevant conflict of interest, although C.E.M. is a Co-Founder of Onegevity.

\section*{References} \addcontentsline{toc}{section}{References}
\begingroup \raggedright \singlespacing \printbibliography[heading=none] \endgroup

% \pagebreak
% \section*{Figures} \addcontentsline{toc}{section}{Figures}
% 
% \begin{figure}[h!] \centering
% \includegraphics[height=.85\textheight,width=\textwidth,keepaspectratio]{figures/HG002-alignment.pdf}
% \caption{
%     Mapping of candidate telomeric PacBio CCS reads from the HG002 dataset.
%     Chromosomes are displayed schematically, centered around the centromere, with only the arms shown to which candidate reads aligned.
%     Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
%     Coordinates are given in bp, relative to the positions of the telomeric tract boundaries.
% }
% \label{fig:hg002_alignment}
% \end{figure}
% \clearpage \pagebreak
% 
% %\addcontentsline{toc}{subsection}{\autoref{fig:hg002_densityplot_q_arm}}
% \begin{figure}[h!] \centering
% \includegraphics[height=\textheight,width=\textwidth,keepaspectratio]{figures/HG002-densityplot-q_arm.pdf}
% \caption{
%     Motif densities at ends of chromosomal \textit{q} arms of the HG002 dataset.
%     Only the arms covered by at least 20 reads are displayed.
%     Shaded boxes span the mapped regions of the genome.
%     Motif densities are plotted as stacked area charts; ribbons surrounding area boundaries represent the 95\% confidence interval of bootstrap.
%     Top four enriched motifs (contributing to at least 0.5\% of the repeat content) are plotted in color; pale tinted areas represent the density of any other motifs and non-repeating sequences (absence of enriched motifs).
%     Absolute genomic coordinates are given in Mbp on the specific reference contigs the reads mapped to (for example, for chr5, reads mapped to the 500 Kbp-long subtelomeric assembly \mbox{5qtel\_1-500K\_1\_12\_12}).
%     Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
% }
% \label{fig:hg002_densityplot_q_arm}
% \end{figure}
% \clearpage \pagebreak
% 
% %\addcontentsline{toc}{subsection}{\autoref{fig:levenshtein_q_arm}}
% \begin{figure}[ht!] \centering
% \includegraphics[height=.9\textheight,width=\textwidth,keepaspectratio]{figures/HG002-levenshtein-densityplots.pdf}
% \caption{
%     Clustering of reads into haplotypes based on relative pairwise Levenshtein distances on four representative chromosomal \textit{q} arms in the HG002 dataset, and densities of top enriched motifs in each haplotype.
%     Genomic coordinates are given in Mbp.
%     Read coverage of each haplotype is annotated above the density plot.
% }
% \label{fig:levenshtein_q_arm}
% \end{figure}
% \clearpage \pagebreak
% 
% \pagebreak
% \section*{Tables} \addcontentsline{toc}{section}{Tables}
% 
% \input{tables/repeatfinder-full.tex}
% \input{tables/HG002-levenshtein-q_arm.tex}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \beginsupplement
% 
% \singlespacing
% \addcontentsline{toc}{section}{Supplemental materials}
% \begin{center}
%     \LARGE{\textbf{Haplotype Diversity and Sequence Heterogeneity of Human Telomeres}}
%     \\~\\
%     \small{Kirill Grigorev, Jonathan Foox \textit{et al.}}
% \section*{Supplemental materials}
% \end{center}
% \doublespacing
% 
% \subsection*{Supplemental methods} \addcontentsline{toc}{subsection}{Supplemental methods} \label{sec:supp_methods}
% 
% To test the presence of non-canonical repeat motifs in datasets generated by technologies other than SMRT,
% we generated four whole-genome Illumina datasets (mean coverage $\sim$104x)
% and three linked-read 10X datasets (mean coverage $\sim$28x)
% for one individual at different timepoints aboard the International Space Station (ISS),
% and one additional linked-read 10X dataset (coverage $\sim$47x) for another individual aboard the ISS.
% From these datasets, candidate telomeric short reads were selected using \textit{Telomerecat} \cite{telomerecat},
% and enriched repeated motifs were discovered \textit{de novo} with the method described in \hyperref[sec:methods]{Materials and Methods}.
% \textit{p}-values were combined with the Mudholkar-George method \cite{george} within each technology (Illumina, 10X),
% and the Bonferroni multiple testing correction was applied
% (note: the Bonferroni correction was applied simultaneously to the \textit{p}-values for the motifs in PacBio reads (\autoref{tab:repeatfinder_full}) and for the motifs in short and linked reads).
% Motifs that were significantly enriched (adjusted \textit{p}-value below the cutoff of 0.05) in the datasets produced by all three technologies (PacBio, Illumina, 10X), with respect to reverse-complemented equivalence, were reported (\autoref{tab:shortread_repeatfinder}).
% 
% \pagebreak
% 
% \subsection*{Supplemental figures} \addcontentsline{toc}{subsection}{Supplemental figures} \label{sec:supp_figs}
% 
% \begin{figure}[ht!] \centering
% \includegraphics[height=.65\textheight,width=\textwidth,keepaspectratio]{figures/HG00X-alignments.pdf}
% \caption{
%     Mapping of candidate telomeric PacBio CCS reads from datasets (A) HG001 and (B) HG005.
%     Chromosomes are displayed schematically, centered around the centromere, with only the arms shown to which candidate reads aligned.
%     Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
%     Coordinates are given in bp, relative to the positions of the telomeric tract boundaries.
% }
% \label{fig:hg00x_alignments}
% \end{figure}
% \clearpage \pagebreak
% 
% \begin{figure}[ht!] \centering
% \includegraphics[height=.9\textheight,width=\textwidth,keepaspectratio]{figures/threemotifp/HG002-densityplot-p_arm-threemotifp.pdf}
% \caption{
%     Densities of top three enriched motifs (contributing to at least 0.5\% of the repeat content) at ends of chromosomal \textit{p} arms of the HG002 dataset.
%     Only the arms covered by at least 20 reads are displayed.
%     Genomic coordinates are given in Mbp.
%     Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
% }
% \label{fig:hg002_densityplot_p_arm}
% \end{figure}
% \clearpage \pagebreak
% 
% \begin{figure}[ht!] \centering
% \includegraphics[height=.95\textheight,width=\textwidth,keepaspectratio]{figures/threemotifp/HG001-densityplots-threemotifp.pdf}
% \caption{
%     Motif densities at ends of chromosomal (A) \textit{p} and (B) \textit{q} arms of the HG001 dataset.
%     Only the arms covered by at least 20 reads are displayed.
%     Genomic coordinates are given in Mbp.
% }
% \label{fig:hg001_densityplots}
% \end{figure}
% \clearpage \pagebreak
% 
% \begin{figure}[ht!] \centering
% \includegraphics[height=.95\textheight,width=\textwidth,keepaspectratio]{figures/threemotifp/HG005-densityplots-threemotifp.pdf}
% \caption{
%     Motif densities at ends of chromosomal (A) \textit{p} and (B) \textit{q} arms of the HG005 dataset.
%     Only the arms covered by at least 20 reads are displayed.
%     Genomic coordinates are given in Mbp.
% }
% \label{fig:hg005_densityplots}
% \end{figure}
% \clearpage \pagebreak
% 
% \begin{figure}[ht!] \centering
% \includegraphics[height=.95\textheight,width=.65\textwidth,keepaspectratio]{figures/entropy.pdf}
% \caption{
%     Distribution of motif entropies in 10 bp windows of candidate PacBio CCS reads aligning to the same chromosomal arms in GIAB datasets HG001, HG002, and HG005.
%     Red solid lines denote the position of the median (0.000 in all three datasets), and red dashed lines denote the 3rd quartile (0.166, 0.074, and 0.211, respectively).
% }
% \label{fig:entropy}
% \end{figure}
% \clearpage \pagebreak
% 
% \subsection*{Supplemental tables} \addcontentsline{toc}{subsection}{Supplemental tables}
% \input{tables/telomeric-read-counts.tex}
% \input{tables/shortread-repeatfinder.tex}
% \input{tables/HG002-haplotype-assignment.tex}

\end{document}
