\documentclass{article}

\usepackage[tmargin=.9in, bmargin=.9in, lmargin=1in, rmargin=1in]{geometry}
\usepackage{bookmark, wrapfig, enumitem, pdflscape, hyphenat}
\usepackage[labelfont=bf]{caption}
\usepackage{lmodern}
\usepackage[sfdefault]{roboto}
\usepackage[T1]{fontenc}
\usepackage{setspace}

\usepackage{graphicx}
    \makeatletter % tex.stackexchange.com/a/28565
    \setlength{\@fptop}{0pt}
    \setlength{\@fpbot}{0pt plus 1fil}
    \makeatother

\usepackage[absolute, overlay]{textpos}
    \setlength{\TPHorizModule}{1mm}
    \setlength{\TPVertModule}{1mm}

\usepackage{xcolor}
    \definecolor{WCM}{RGB}{172,31,44} % AC1E2C

\usepackage{hyperref} \hypersetup{
    colorlinks=true, linkcolor={blue!65!black},
    citecolor={blue!65!black}, urlcolor={blue!50!black},
    pdfpagelayout=OneColumn, pdfstartview={XYZ null null 1.25},
    bookmarksnumbered=true, bookmarksopen=true, bookmarksopenlevel=3
}

\usepackage[backend=bibtex, style=authoryear]{biblatex}
    \addbibresource{references.bib}
    % https://tex.stackexchange.com/a/134065 :
    \renewcommand*{\nameyeardelim}{\addcomma\space}
    % https://tex.stackexchange.com/a/404787 :
    \usepackage{xpatch}
    \DeclareNameAlias{sortname}{last-first} % general changes
    \renewcommand*{\bibinitdelim}{}
    \renewbibmacro*{in:}{\iffieldequalstr{entrytype}{inproceedings}{\printtext{\bibstring{in}\addspace}}{}}
    \csletcs{abx@macro@publisher+location+date@orig}{abx@macro@publisher+location+date} % changes for "book"
    \renewbibmacro*{publisher+location+date}{\printtext[parens]{\usebibmacro{publisher+location+date@orig}}}
    \DeclareFieldFormat[book]{title}{#1\printunit{\addspace}}
    \DeclareFieldFormat[inproceedings]{title}{#1\isdot} % changes for "inproceedings"
    \DeclareFieldFormat{booktitle}{#1\addcomma}
    \xpatchbibmacro{byeditor+others}
        {\usebibmacro{byeditor+othersstrg} \setunit{\addspace} \printnames[byeditor]{editor} \clearname{editor}}
        {\printnames[byeditor]{editor} \clearname{editor} \addcomma\addspace \bibstring{editor} \setunit{\addspace}}{}{}
    \DeclareFieldFormat[article]{title}{#1} % changes in "article"
    \DeclareFieldFormat[article]{journaltitle}{#1\isdot}
    \DeclareFieldFormat[article]{volume}{\textit{#1}}
    \DeclareFieldFormat[article]{pages}{#1}
    % https://en.wikibooks.org/wiki/LaTeX/Macros#New_commands
    \newcommand{\citep}[1]{(\cite{#1})}

\newcommand{\beginsupplement}{
% bytesizebio.net/2013/03/11/adding-supplementary-tables-and-figures-in-latex
    %\newpage
    %\setcounter{page}{1}
    %\renewcommand{\thepage}{S-\arabic{page}}
    \pagenumbering{gobble}
    \setcounter{table}{0}
    \renewcommand{\thetable}{S\arabic{table}}
    \setcounter{figure}{0}
    \renewcommand{\thefigure}{S\arabic{figure}}
 }
\begin{document}

\beginsupplement

\section*{Supplemental Information}

\subsection*{Supplemental figures} \addcontentsline{toc}{subsection}{Supplemental figures} \label{sec:supp_figs}

\begin{figure}[ht!] \centering
\includegraphics[height=.65\textheight,width=\textwidth,keepaspectratio]{figures/HG00X-alignments.pdf}
\caption{
    Mapping of candidate telomeric PacBio CCS reads from datasets (A) HG001 and (B) HG005.
    Chromosomes are displayed schematically, centered around the centromere, with only the arms shown to which candidate reads aligned.
    Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
    Coordinates are given in bp, relative to the positions of the telomeric tract boundaries.
    Relates to: \textbf{Figure 1}.
}
\label{fig:hg00x_alignments}
\end{figure}
\clearpage \pagebreak

\begin{figure}[ht!] \centering
\includegraphics[height=.9\textheight,width=\textwidth,keepaspectratio]{figures/threemotifp/HG002-densityplot-p_arm-threemotifp.pdf}
\caption{
    Densities of top three enriched motifs (contributing to at least 0.5\% of the repeat content) at ends of chromosomal \textit{p} arms of the HG002 dataset.
    Only the arms covered by at least 20 reads are displayed.
    Genomic coordinates are given in Mbp.
    Vertical red dashed lines denote the position of the boundary of the annotated telomeric tract.
    Relates to: \textbf{Figure 2}, \textbf{Table 1}.
}
\label{fig:hg002_densityplot_p_arm}
\end{figure}
\clearpage \pagebreak

\begin{figure}[ht!] \centering
\includegraphics[height=.95\textheight,width=\textwidth,keepaspectratio]{figures/threemotifp/HG001-densityplots-threemotifp.pdf}
\caption{
    Motif densities at ends of chromosomal (A) \textit{p} and (B) \textit{q} arms of the HG001 dataset.
    Only the arms covered by at least 20 reads are displayed.
    Genomic coordinates are given in Mbp.
    Relates to: \textbf{Figure 2}, \textbf{Table 1}.
}
\label{fig:hg001_densityplots}
\end{figure}
\clearpage \pagebreak

\begin{figure}[ht!] \centering
\includegraphics[height=.95\textheight,width=\textwidth,keepaspectratio]{figures/threemotifp/HG005-densityplots-threemotifp.pdf}
\caption{
    Motif densities at ends of chromosomal (A) \textit{p} and (B) \textit{q} arms of the HG005 dataset.
    Only the arms covered by at least 20 reads are displayed.
    Genomic coordinates are given in Mbp.
    Relates to: \textbf{Figure 2}, \textbf{Table 1}.
}
\label{fig:hg005_densityplots}
\end{figure}
\clearpage \pagebreak

\begin{figure}[ht!] \centering
\includegraphics[height=.95\textheight,width=.65\textwidth,keepaspectratio]{figures/entropy.pdf}
\caption{
    Distribution of motif entropies in 10 bp windows of candidate PacBio CCS reads aligning to the same chromosomal arms in GIAB datasets HG001, HG002, and HG005.
    Red solid lines denote the position of the median (0.000 in all three datasets), and red dashed lines denote the 3rd quartile (0.166, 0.074, and 0.211, respectively).
    Relates to: \textbf{STAR Methods, Evaluation of sequence concordance in telomeric long reads}.
}
\label{fig:entropy}
\end{figure}
\clearpage \pagebreak

\subsection*{Supplemental tables} \addcontentsline{toc}{subsection}{Supplemental tables}
\input{tables/telomeric-read-counts.tex}
\input{tables/shortread-repeatfinder.tex}
\input{tables/HG002-haplotype-assignment.tex}

\end{document}
