{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import re\n",
    "from matplotlib.pyplot import subplots, style, rc\n",
    "from tqdm import tqdm\n",
    "from venn import venn, pseudovenn\n",
    "from collections import defaultdict\n",
    "from itertools import count, islice\n",
    "from functools import lru_cache\n",
    "from pysam import AlignmentFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT_TO_TRIO = {\n",
    "    \"HG001\": \"NA12878\",\n",
    "    \"HG002\": \"AshkenazimTrio\", \"HG003\": \"AshkenazimTrio\", \"HG004\": \"AshkenazimTrio\",\n",
    "    \"HG005\": \"ChineseTrio\", \"HG006\": \"ChineseTrio\", \"HG007\": \"ChineseTrio\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecx = pd.read_csv(\"assets/hg38ext.fa.ecx\", sep=\"\\t\", skiprows=1, escapechar=\"#\") \\\n",
    "    .query(\"flag==16384\").query(\"blacklist=='-'\") \\\n",
    "    .drop(columns=[\"entry\", \"pos+1\", \"main_rname\", \"flag\", \"link\", \"blacklist\", \"class\"])\n",
    "\n",
    "ecx[\"sorter\"] = ecx[\"chromosome\"].apply(lambda c: 999 if c == \"chrX\" else int(c[3:]))\n",
    "ecx = ecx.sort_values(by=\"sorter\").drop(columns=\"sorter\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_bam(filename):\n",
    "    p_arm, q_arm = defaultdict(list), defaultdict(list)\n",
    "    with AlignmentFile(filename) as bam:\n",
    "        for entry in bam:\n",
    "            if (entry.flag & 0x4000 == 0x4000) and (entry.seq is not None) and (entry.flag & 3844 == 0): # tract_anchor\n",
    "                entry_stats = [\n",
    "                    entry.reference_start, entry.query_alignment_start,\n",
    "                    entry.reference_end, entry.query_alignment_end,\n",
    "                    len(entry.seq)\n",
    "                ]\n",
    "                if entry.flag & 0x8000 == 0x8000: # is_q\n",
    "                    q_arm[entry.reference_name].append(entry_stats)\n",
    "                else:\n",
    "                    p_arm[entry.reference_name].append(entry_stats)\n",
    "    return p_arm, q_arm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def arm_to_counts(arm, name):\n",
    "    return pd.Series({rname: len(reads) for rname, reads in arm.items()}, name=name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_counts, q_counts = [], []\n",
    "\n",
    "for subject, trio in SUBJECT_TO_TRIO.items():\n",
    "    p_arm, q_arm = load_bam(f\"PacBio/{trio}/{subject}/tailpuller.bam\")\n",
    "    p_counts.append(arm_to_counts(p_arm, subject))\n",
    "    q_counts.append(arm_to_counts(q_arm, subject))\n",
    "\n",
    "raw_p_counts = pd.concat(p_counts, axis=1, sort=False).fillna(0).astype(int)\n",
    "raw_p_counts.index.name = \"rname\"\n",
    "raw_p_counts.reset_index(inplace=True)\n",
    "\n",
    "raw_q_counts = pd.concat(q_counts, axis=1, sort=False).fillna(0).astype(int)\n",
    "raw_q_counts.index.name = \"rname\"\n",
    "raw_q_counts.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>HG001</th>\n",
       "      <th>HG002</th>\n",
       "      <th>HG003</th>\n",
       "      <th>HG004</th>\n",
       "      <th>HG005</th>\n",
       "      <th>HG006</th>\n",
       "      <th>HG007</th>\n",
       "      <th>total</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arm</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>rname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">p</th>\n",
       "      <th>chr10</th>\n",
       "      <th>chr10</th>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12</th>\n",
       "      <th>chr12</th>\n",
       "      <td>10</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr16</th>\n",
       "      <th>chr16</th>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr17</th>\n",
       "      <th>17ptel_1_500K_1_12_12</th>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr18</th>\n",
       "      <th>chr18</th>\n",
       "      <td>34</td>\n",
       "      <td>94</td>\n",
       "      <td>46</td>\n",
       "      <td>22</td>\n",
       "      <td>21</td>\n",
       "      <td>36</td>\n",
       "      <td>68</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr19</th>\n",
       "      <th>19ptel_1-500K_1_12_12</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2</th>\n",
       "      <th>chr2</th>\n",
       "      <td>10</td>\n",
       "      <td>63</td>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5</th>\n",
       "      <th>chr5</th>\n",
       "      <td>18</td>\n",
       "      <td>61</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr8</th>\n",
       "      <th>chr8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9</th>\n",
       "      <th>chr9</th>\n",
       "      <td>9</td>\n",
       "      <td>26</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>59</td>\n",
       "      <td>73</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <th></th>\n",
       "      <td>91</td>\n",
       "      <td>295</td>\n",
       "      <td>113</td>\n",
       "      <td>111</td>\n",
       "      <td>53</td>\n",
       "      <td>144</td>\n",
       "      <td>209</td>\n",
       "      <td>1016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"19\" valign=\"top\">q</th>\n",
       "      <th>chr10</th>\n",
       "      <th>10qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11</th>\n",
       "      <th>chr11</th>\n",
       "      <td>24</td>\n",
       "      <td>63</td>\n",
       "      <td>35</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>23</td>\n",
       "      <td>50</td>\n",
       "      <td>231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12</th>\n",
       "      <th>chr12</th>\n",
       "      <td>19</td>\n",
       "      <td>58</td>\n",
       "      <td>46</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr14</th>\n",
       "      <th>14qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>14</td>\n",
       "      <td>67</td>\n",
       "      <td>31</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15</th>\n",
       "      <th>chr15</th>\n",
       "      <td>51</td>\n",
       "      <td>36</td>\n",
       "      <td>26</td>\n",
       "      <td>31</td>\n",
       "      <td>41</td>\n",
       "      <td>80</td>\n",
       "      <td>54</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr16</th>\n",
       "      <th>16qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr17</th>\n",
       "      <th>17qtel_1-500K_1_12_12v2_rc</th>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr18</th>\n",
       "      <th>18qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>11</td>\n",
       "      <td>57</td>\n",
       "      <td>21</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr19</th>\n",
       "      <th>chr19</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "      <td>17</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2</th>\n",
       "      <th>2qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21</th>\n",
       "      <th>chr21</th>\n",
       "      <td>54</td>\n",
       "      <td>157</td>\n",
       "      <td>52</td>\n",
       "      <td>37</td>\n",
       "      <td>52</td>\n",
       "      <td>40</td>\n",
       "      <td>42</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22</th>\n",
       "      <th>chr22</th>\n",
       "      <td>5</td>\n",
       "      <td>138</td>\n",
       "      <td>33</td>\n",
       "      <td>52</td>\n",
       "      <td>10</td>\n",
       "      <td>66</td>\n",
       "      <td>57</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3</th>\n",
       "      <th>chr3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5</th>\n",
       "      <th>5qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>68</td>\n",
       "      <td>154</td>\n",
       "      <td>104</td>\n",
       "      <td>75</td>\n",
       "      <td>35</td>\n",
       "      <td>60</td>\n",
       "      <td>102</td>\n",
       "      <td>598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr6</th>\n",
       "      <th>6qtel_1-500K_1_12_12_rc</th>\n",
       "      <td>45</td>\n",
       "      <td>121</td>\n",
       "      <td>47</td>\n",
       "      <td>45</td>\n",
       "      <td>39</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7</th>\n",
       "      <th>chr7</th>\n",
       "      <td>19</td>\n",
       "      <td>72</td>\n",
       "      <td>27</td>\n",
       "      <td>21</td>\n",
       "      <td>16</td>\n",
       "      <td>23</td>\n",
       "      <td>25</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr8</th>\n",
       "      <th>chr8</th>\n",
       "      <td>22</td>\n",
       "      <td>74</td>\n",
       "      <td>32</td>\n",
       "      <td>22</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>38</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX</th>\n",
       "      <th>chrX</th>\n",
       "      <td>43</td>\n",
       "      <td>107</td>\n",
       "      <td>22</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>23</td>\n",
       "      <td>36</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <th></th>\n",
       "      <td>384</td>\n",
       "      <td>1119</td>\n",
       "      <td>496</td>\n",
       "      <td>388</td>\n",
       "      <td>329</td>\n",
       "      <td>487</td>\n",
       "      <td>526</td>\n",
       "      <td>3729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>total</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <td>475</td>\n",
       "      <td>1414</td>\n",
       "      <td>609</td>\n",
       "      <td>499</td>\n",
       "      <td>382</td>\n",
       "      <td>631</td>\n",
       "      <td>735</td>\n",
       "      <td>4745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             HG001  HG002  HG003  HG004  \\\n",
       "arm   chromosome rname                                                    \n",
       "p     chr10      chr10                           2     12      1      5   \n",
       "      chr12      chr12                          10     21     16      8   \n",
       "      chr16      chr16                           8      2      0     12   \n",
       "      chr17      17ptel_1_500K_1_12_12           0      9      1      7   \n",
       "      chr18      chr18                          34     94     46     22   \n",
       "      chr19      19ptel_1-500K_1_12_12           0      7      0      0   \n",
       "      chr2       chr2                           10     63     18     25   \n",
       "      chr5       chr5                           18     61     14     12   \n",
       "      chr8       chr8                            0      0      0      4   \n",
       "      chr9       chr9                            9     26     17     16   \n",
       "      total                                     91    295    113    111   \n",
       "q     chr10      10qtel_1-500K_1_12_12_rc        0      2      0      0   \n",
       "      chr11      chr11                          24     63     35     20   \n",
       "      chr12      chr12                          19     58     46     27   \n",
       "      chr14      14qtel_1-500K_1_12_12_rc       14     67     31     11   \n",
       "      chr15      chr15                          51     36     26     31   \n",
       "      chr16      16qtel_1-500K_1_12_12_rc        0      6      6      8   \n",
       "      chr17      17qtel_1-500K_1_12_12v2_rc      1      6      1      2   \n",
       "      chr18      18qtel_1-500K_1_12_12_rc       11     57     21     11   \n",
       "      chr19      chr19                           8      1     13      7   \n",
       "      chr2       2qtel_1-500K_1_12_12_rc         0      0      0      0   \n",
       "      chr21      chr21                          54    157     52     37   \n",
       "      chr22      chr22                           5    138     33     52   \n",
       "      chr3       chr3                            0      0      0      3   \n",
       "      chr5       5qtel_1-500K_1_12_12_rc        68    154    104     75   \n",
       "      chr6       6qtel_1-500K_1_12_12_rc        45    121     47     45   \n",
       "      chr7       chr7                           19     72     27     21   \n",
       "      chr8       chr8                           22     74     32     22   \n",
       "      chrX       chrX                           43    107     22     16   \n",
       "      total                                    384   1119    496    388   \n",
       "total                                          475   1414    609    499   \n",
       "\n",
       "                                             HG005  HG006  HG007  total  \n",
       "arm   chromosome rname                                                   \n",
       "p     chr10      chr10                           2      5      4     31  \n",
       "      chr12      chr12                           6     16      6     83  \n",
       "      chr16      chr16                           0      3      0     25  \n",
       "      chr17      17ptel_1_500K_1_12_12           1      3      8     29  \n",
       "      chr18      chr18                          21     36     68    321  \n",
       "      chr19      19ptel_1-500K_1_12_12           2      0      5     14  \n",
       "      chr2       chr2                            6     12     11    145  \n",
       "      chr5       chr5                           11     10     31    157  \n",
       "      chr8       chr8                            0      0      3      7  \n",
       "      chr9       chr9                            4     59     73    204  \n",
       "      total                                     53    144    209   1016  \n",
       "q     chr10      10qtel_1-500K_1_12_12_rc        0     15      0     17  \n",
       "      chr11      chr11                          16     23     50    231  \n",
       "      chr12      chr12                          20     18     15    203  \n",
       "      chr14      14qtel_1-500K_1_12_12_rc        9     13     13    158  \n",
       "      chr15      chr15                          41     80     54    319  \n",
       "      chr16      16qtel_1-500K_1_12_12_rc        0      2      8     30  \n",
       "      chr17      17qtel_1-500K_1_12_12v2_rc      1      8      2     21  \n",
       "      chr18      18qtel_1-500K_1_12_12_rc        9     18     18    145  \n",
       "      chr19      chr19                          24     27     17     97  \n",
       "      chr2       2qtel_1-500K_1_12_12_rc         1      0      0      1  \n",
       "      chr21      chr21                          52     40     42    434  \n",
       "      chr22      chr22                          10     66     57    361  \n",
       "      chr3       chr3                            0      1      0      4  \n",
       "      chr5       5qtel_1-500K_1_12_12_rc        35     60    102    598  \n",
       "      chr6       6qtel_1-500K_1_12_12_rc        39     49     49    395  \n",
       "      chr7       chr7                           16     23     25    203  \n",
       "      chr8       chr8                           20     21     38    229  \n",
       "      chrX       chrX                           36     23     36    283  \n",
       "      total                                    329    487    526   3729  \n",
       "total                                          382    631    735   4745  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_counts = pd.merge(raw_p_counts, ecx.loc[ecx[\"prime\"]==5, [\"rname\", \"chromosome\"]]).sort_values(by=\"chromosome\")\n",
    "p_counts[\"arm\"] = \"p\"\n",
    "p_counts = p_counts.set_index([\"arm\", \"chromosome\", \"rname\"])\n",
    "\n",
    "q_counts = pd.merge(raw_q_counts, ecx.loc[ecx[\"prime\"]==3, [\"rname\", \"chromosome\"]]).sort_values(by=\"chromosome\")\n",
    "q_counts[\"arm\"] = \"q\"\n",
    "q_counts = q_counts.set_index([\"arm\", \"chromosome\", \"rname\"])\n",
    "\n",
    "assert len(p_counts) == len(raw_p_counts)\n",
    "assert len(q_counts) == len(raw_q_counts)\n",
    "\n",
    "p_counts.loc[(\"p\", \"total\", \"\")] = p_counts.sum()\n",
    "q_counts.loc[(\"q\", \"total\", \"\")] = q_counts.sum()\n",
    "\n",
    "counts = pd.concat([p_counts, q_counts])\n",
    "counts.loc[(\"total\", \"\", \"\")] = p_counts.loc[(\"p\", \"total\", \"\")] + q_counts.loc[(\"q\", \"total\", \"\")]\n",
    "\n",
    "counts[\"total\"] = counts.sum(axis=1)\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts.to_csv(\"Table-S1-coverage.tsv\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
